wicket-core/src/main/java/org/apache/wicket/pageStore/DiskPageStore.java

```diff
@@ -0,0 +1,717 @@
+		this.folders = new NestedFolders(new File(fileStoreFolder, applicationName + "-filestore"));
+		this.maxSizePerSession = Args.notNull(maxSizePerSession, "maxSizePerSession");
+		this.serializer = serializer; // optional
+
+		this.diskDatas = new ConcurrentHashMap<>();
+
+		try
+		{
+			if (folders.getBase().exists() || folders.getBase().mkdirs())
+			{
+				loadIndex();
+			}
+			else
+			{
+				log.warn("Cannot create file store folder for some reason.");
+			}
+		}
+		catch (SecurityException e)
+		{
+			throw new WicketRuntimeException(
+				"SecurityException occurred while creating DiskPageStore. Consider using a non-disk based IPageStore implementation. "
+					+ "See org.apache.wicket.Application.setPageManagerProvider(IPageManagerProvider)",
+				e);
+		}
+
+		if (DISK_STORES.containsKey(applicationName))
+		{
+			throw new IllegalStateException(
+				"Store for application with key '" + applicationName + "' already exists.");
+		}
+		DISK_STORES.put(applicationName, this);
+	}
+
+	/**
+	 * Versioning is supported if a serializer was provided to the constructor.
+	 */
+	@Override
+	public boolean supportsVersioning()
+	{
+		return serializer != null;
+	}
+
+	@Override
+	public void destroy()
+	{
+		log.debug("Destroying...");
+		saveIndex();
+
+		DISK_STORES.remove(applicationName);
+
+		log.debug("Destroyed.");
+	}
+
+	@Override
+	public boolean canBeAsynchronous(IPageContext context)
+	{
+		// session attribute must be added here *before* any asynchronous calls
+		// when session is no longer available
+		getSessionAttribute(context, true);
+
+		return true;
+	}
+
+	@Override
+	public IManageablePage getPage(IPageContext context, int id)
+	{
+		IManageablePage page = null;
+
+		DiskData diskData = getDiskData(context, false);
+		if (diskData != null)
+		{
+			byte[] data = diskData.loadPage(id);
+			if (data != null)
+			{
+				if (serializer == null)
+				{
+					page = new SerializedPage(id, "unknown", data);
+				}
+				else
+				{
+					page = (IManageablePage)serializer.deserialize(data);
+				}
+
+				if (log.isDebugEnabled())
+				{
+					log.debug("Returning page with id '{}' in session with id '{}'", id, context.getSessionId());
+				}
+			}
+		}
+
+		return page;
+	}
+
+	@Override
+	public void removePage(IPageContext context, IManageablePage page)
+	{
+		DiskData diskData = getDiskData(context, false);
+		if (diskData != null)
+		{
+			if (log.isDebugEnabled())
+			{
+				log.debug("Removing page with id '{}' in session with id '{}'", page.getPageId(),
+					context.getSessionId());
+			}
+			diskData.removeData(page.getPageId());
+		}
+	}
+
+	@Override
+	public void removeAllPages(IPageContext context)
+	{
+		DiskData diskData = getDiskData(context, false);
+		if (diskData != null)
+		{
+			removeDiskData(diskData);
+		}
+	}
+
+	protected void removeDiskData(DiskData diskData)
+	{
+		synchronized (diskDatas)
+		{
+			diskDatas.remove(diskData.sessionIdentifier);
```

Comment:
Yes, the wording is difficult here: The sessionId isn't always equal to the 'session identifier' (see dicussion above), the former might change (e.g. the container changes it), but the latter must stay the same so that data stored in a separate storage (disk, files, cassandra) can still be identified.